src/maxdiffusion/loaders/lora_pipeline.py

```diff
@@ -0,0 +1,280 @@
+            `default_{i}` where i is the total number of adapters being loaded.
+    """
+    # if a dict is passed, copy it instead of modifying it inplace
+    if isinstance(pretrained_model_name_or_path_or_dict, dict):
+      pretrained_model_name_or_path_or_dict = pretrained_model_name_or_path_or_dict.copy()
+
+    state_dict, network_alphas = self.lora_state_dict(pretrained_model_name_or_path_or_dict, **kwargs)
+    is_correct_format = all("lora" in key or "dora_scale" in key for key in state_dict.keys())
+    if not is_correct_format:
+      raise ValueError("Invalid LoRA checkpoint.")
+
+    params, rank, network_alphas = self.load_lora(
+        state_dict,
+        network_alphas=network_alphas,
+        params=params,
+        adapter_name=adapter_name,
+        _pipeline=self,
+    )
+    return params, rank, network_alphas
+
+  @classmethod
+  def _get_lora_layer(cls, module_path, module, rank, network_alphas):
+    is_conv = any("conv" in str_ for str_ in module_path)
+    network_alpha = network_alphas.get(module_path, None)
+    if is_conv:
+      lora_module = LoRAConv2DLayer(
+          out_features=module.features,
+          rank=rank,
+          network_alpha=network_alpha,
+          kernel_size=module.kernel_size,
+          strides=module.strides,
+          padding=module.padding,
+          input_dilation=module.input_dilation,
+          kernel_dilation=module.kernel_dilation,
+          feature_group_count=module.feature_group_count,
+          dtype=module.dtype,
+          weights_dtype=module.param_dtype,
+          precision=module.precision,
+          name="lora",
+      )
+    else:
+      lora_module = LoRALinearLayer(
+          out_features=module.features,
+          rank=rank,
+          network_alpha=network_alpha,
+          dtype=module.dtype,
+          weights_dtype=module.param_dtype,
+          precision=module.precision,
+          name="lora",
+      )
+    return lora_module
+
+  def rename_for_interceptor(params_keys, network_alphas):
+    new_params_keys = []
+    for layer_lora in params_keys:
+      if "lora" in layer_lora:
+        new_layer_lora = layer_lora[: layer_lora.index("lora")]
+        if new_layer_lora not in new_params_keys:
+          new_params_keys.append(new_layer_lora)
+          network_alpha = network_alphas[layer_lora]
+          del network_alphas[layer_lora]
+          network_alphas[new_layer_lora] = network_alpha
+    return new_params_keys, network_alphas
+
+  @classmethod
+  def make_lora_interceptor(cls, params, rank, network_alphas):
+    # Only unet interceptor supported for now.
+    unet_lora_keys = flax.traverse_util.flatten_dict(params["unet"]).keys()
+    unet_lora_keys, network_alphas = cls.rename_for_interceptor(unet_lora_keys, network_alphas)
+
+    def _intercept(next_fn, args, kwargs, context):
+      mod = context.module
+      while mod is not None:
```

Comment:
Can you explain what's happening with this while loop here?

Reply:
This is to prevent recursion. Because LoRA layers are being added dynamically, it can cause an endless loop as they can be "re-intercepted".

Reply:
Ah. For my understanding of how these interceptors work, say originally we had something like `Linear -> Non-linearity -> Linear`. Say we get a LoRa that's added in dynamically, so we now have  `Linear -> LoRa -> Non-linearity -> Linear`. Is this trying to prevent the case where the LoRa layer gets intercepted and another LoRa layer gets added? Does the check on Line 149 not catch this?

Reply:
Yes that is correct. Line 149 would catch this, but I added this for completeness, future proofing.

Reply:
Got it - it might be useful to add a comment explaining what's happening here, but that can be done in a follow-up PR.